src/EditorFeatures/VisualBasic/AutomaticEndConstructCorrection/AbstractCorrector.vb

```diff
@@ -0,0 +1,209 @@
+            _buffer = subjectBuffer
+            _waitIndicator = waitIndicator
+
+            _session = New Session(subjectBuffer, AddressOf ShouldReplaceText)
+            _previousDocument = Nothing
+            _referencingViews = 0
+        End Sub
+
+        Protected MustOverride Function IsAllowableWordAtIndex(lineText As String, wordStartIndex As Integer, wordLength As Integer) As Boolean
+        Protected MustOverride Function TryGetValidTokens(wordStartIndex As Integer, ByRef startToken As SyntaxToken, ByRef endToken As SyntaxToken, cancellationToken As CancellationToken) As Boolean
+        Protected MustOverride Function ShouldReplaceText(text As String) As Boolean
+
+        Protected ReadOnly Property PreviousDocument As Document
+            Get
+                Return _previousDocument
+            End Get
+        End Property
+
+        Public Sub Connect() Implements ICorrector.Connect
+            If _referencingViews = 0 Then
+                AddHandler _buffer.Changing, AddressOf OnTextBufferChanging
+                AddHandler _buffer.Changed, AddressOf OnTextBufferChanged
+            End If
+
+            _referencingViews = _referencingViews + 1
+        End Sub
+
+        Public Sub Disconnect() Implements ICorrector.Disconnect
+            If _referencingViews = 1 Then
+                RemoveHandler _buffer.Changed, AddressOf OnTextBufferChanged
+                RemoveHandler _buffer.Changing, AddressOf OnTextBufferChanging
+            End If
+
+            _referencingViews = Math.Max(_referencingViews - 1, 0)
+        End Sub
+
+        Public ReadOnly Property IsDisconnected As Boolean Implements ICorrector.IsDisconnected
+            Get
+                Return _referencingViews = 0
+            End Get
+        End Property
+
+        Protected Function CreateTrackingSpans(startToken As SyntaxToken, endToken As SyntaxToken, snapshot As ITextSnapshot) As IEnumerable(Of ITrackingSpan)
+            Return {CreateTrackingSpan(startToken, snapshot), New LetterOnlyTrackingSpan(endToken.Span.ToSnapshotSpan(snapshot))}
+        End Function
+
+        Private Function CreateTrackingSpan(token As SyntaxToken, snapshot As ITextSnapshot) As ITrackingSpan
+            Return New LetterOnlyTrackingSpan(token.Span.ToSnapshotSpan(snapshot))
+        End Function
+
+        Private Sub OnTextBufferChanging(sender As Object, e As TextContentChangingEventArgs)
+            If Me._session.Alive Then
+                _previousDocument = Nothing
+                Return
+            End If
+
+            ' try holding onto previous Document so that we can use it when we diff syntax tree
+            _previousDocument = e.Before.GetOpenDocumentInCurrentContextWithChanges()
+        End Sub
+
+        Private Sub OnTextBufferChanged(sender As Object, e As TextContentChangedEventArgs)
+            _waitIndicator.Wait(
+                "IntelliSense",
```

Comment:
Not your fault, but any chance you can find a better localized string to use here? This is just a bug.